[WIP] Replace arg --lang with --audio-corpus and --language-model

mpuels opened the pull request:
This is the (work in progress) development branch corresponding to #13, hence I don't expect it to be merged, until it is deemed finished.

The first commit refactors `speech_sentences_{de,en}.py` into the scripts `speech_sentences.py` and `speech_train_punkt_tokenizer.py`.

I tested the new scripts on the following corpora:

- cornell_movie_dialogs
- europarl_de
- europarl_en
- parole_de

Unfortunately I wasn't sure where to download the corpora:

- web_questions
- yahoo_answers

Wouldn't it be nice if the repo contained a script `speech_pull_text_corpus.py` that would fully automatically download text corpora? E.g. `./speech_pull_text_corpus.py web_questions`.

Finally I have a question regarding `speech_sentences_de.py`: What is the rationale behind first training the Punkt tokenizer on the Parole corpus and then later applying the same tokenizer to the text data it has been trained on?

mpuels committed:
Replace speech_sentences_{de,en}.py with speech_sentences.py

Furthermore put training of Punkt tokenizer into own script
speech_train_punkt_tokenizer.py.

mpuels committed:
Change arguments of speech_build_lm.py

mpuels commented:
In the script `speech_build_lm.py` I've removed the argument `--lang`. Instead the script now expects the argument `language_model` and a non-empty list of `text_corpus`.

Note: For now I've removed the code that reads the prompts from `transcripts_*.csv` to put them into `train_all.txt`. I don't want to change the class `Transcripts(lang="de")` yet. But I'm wondering: is it good practice to include the prompts into the language model? Or is this considered bad practice, just as it is considered bad practice to have the same speaker in the training set and test set?

gooofy commented:
> The first commit refactors speech_sentences_{de,en}.py into the scripts
> speech_sentences.py and speech_train_punkt_tokenizer.py.

cool!

> Unfortunately I wasn't sure where to download the corpora:
>
> web_questions

https://nlp.stanford.edu/software/sempre/

> yahoo_answers

https://cogcomp.org/page/resource_view/89

> Wouldn't it be nice if the repo contained a script speech_pull_text_corpus.py
> that would fully automatically download text corpora? E.g.
> ./speech_pull_text_corpus.py web_questions.

sounds good to me although the download URLs and mechanisms may change over
time so we will have to put some maintenance effort into that script if we want
to keep it useful.

that said, we could definitely define a place in our src/ directory structure 
where text corpora should reside and what their files/directories should be
called along with download instructions in the README so users may download
them manually and put them there
 
> Finally I have a question regarding speech_sentences_de.py: What is the
> rationale behind first training the Punkt tokenizer on the Parole corpus and
> then later applying the same tokenizer to the text data it has been trained on?

I am no expert on the Punkt tokenizer, so feel free to correct me if I am 
wrong: From what I understood the punkt tokenizer uses an unsupervised learning
approach, i.e. you just feed it a lot of text and it will figure out
sentence/token boundaries on its own. Once the model is built it can then be
used to segment even its own training data.

gooofy commented:
> In the script speech_build_lm.py I've removed the argument --lang. Instead the
> script now expects the argument language_model and a non-empty list of
> text_corpus.

excellent! :)

> Note: For now I've removed the code that reads the prompts from
> transcripts_*.csv to put them into train_all.txt. I don't want to change the
> class Transcripts(lang="de") yet.

I think we should move that code to a separate tool that generates just text file
(something along the lines of dst/text-corpora/prompts-de.txt maybe?) which can then
be used for training language models in addition to the other text corpora if the
user decides to do so.

> But I'm wondering: is it good practice to
> include the prompts into the language model? Or is this considered bad
> practice, just as it is considered bad practice to have the same speaker in the
> training set and test set?

yeah, I can see why some people would consider that cheating ;) - on the other 
hand my reasoning was that the impact is pretty low given the huge amount of
training material coming from the other training corpora and for my voice
assistant applications it made a lot of sense to have typical voice commands as
part of the training data of the language model.

anyway, if we follow the approach I outlined above the user can decide whether 
he or she wants the prompts to be a part of the lm training data or not as this
would be completely
optional.

mpuels commented:
1. **Script that downloads text corpora** Instead of writing a script I've added links to text corpora to README.md (https://github.com/gooofy/speech/pull/16). That's not as convenient as a script, but better than nothing :) But why should the corpora be placed in `src/`? I thought the user can put them wherevery she likes and just update the corresponding variable in `.speechrc`?!

2. **Training of Punkt tokenizer** You're right, the Punkt tokenizer uses an unsupervised method to learn sentense boundaries, so I guess it makes sense to train it on and apply it to the same corpus.

3. **Let `speech_sentences.py` extract prompts from VoxForge corpus** What do you think if `speech_sentences.py` would recognize the argument `voxforge-de-prompts`, which would extract the prompts and put them into `dst/text-corpora/voxforge-de-prompts.txt`? I've updated the proposed workflow in the first post of issue https://github.com/gooofy/speech/issues/13#issue-308030088 accordingly.

mpuels committed:
Move extracted sentences, language models, and tokenizer to other dir

- extracted sentences are now written to `data/dst/text-corpora/`
- computed language models are now written to `data/dst/lm/`
- tokenizers are now written to `data/dst/tokenizers/`

mpuels committed:
Split German transcripts_*.csv according to speech corpus

mpuels committed:
speech_audio_scan.py distinguishes between corpora now

mpuels commented:
I've rewritten `speech_audio_scan.py` as described here: https://github.com/gooofy/speech/issues/13#issuecomment-378556605

To that end I had to split the contents of `data/src/speech/de/` into `data/src/speech/voxforge_de/` and `data/src/speech/gspv2/`. I wrote the script [split_trancsripts_csv.bash](https://github.com/gooofy/speech/pull/15/commits/b8f0b29e82afb8a6529e45235f0d19fa6817ca27#diff-cbaaf4f4a35093664693910ecbdb9dca) to perform the split. If we agree on the split, we can remove the script.

But there's a thing I'm not sure about: Did the original `speech_audio_scan.py` process files from the gspv2 corpus? When I run the original script, it does not 

1. process wav files
2. update `transcripts_*.csv`

for the gspv2 corpus. Is this behaviour correct? Or am I missing something?

gooofy commented:
> To that end I had to split the contents of data/src/speech/de/ into
> data/src/speech/voxforge_de/ and data/src/speech/gspv2/. I wrote the script
> split_trancsripts_csv.bash to perform the split. If we agree on the split, we
> can remove the script.

excellent work!

> But there's a thing I'm not sure about: Did the original speech_audio_scan.py
> process files from the gspv2 corpus? When I run the original script, it does
> not process wav files update transcripts_*.csv for the gspv2 corpus. Is this
> behaviour correct? Or am I missing something?

ah, you're starting to discover more and more of the dark secrets of this project ;)

actually this is something I had in the back of my mind for some time now but 
never investigated deeper: at some point I ran into trouble with the 
voxforge-inherited tarball/directory naming convention. AFAIR this lead
to the problem that scripts either ignore or complain about the gspv2* files.

For me, that was not a problem as I had already added them to the .csv db files
at that point, so I added that issue to my some day / maybe TODO-list. But I guess
we will have to investigate this a bit deeper at this point :-/

mpuels committed:
Add args gspv2 and voxforge_de to speech_sentences.py

mpuels commented:
**Convert gspv2 to VoxForge format**

I have taken your script `gspv2_to_vf.py` and have integrated into this git repo (but not commited it yet). The part that copies the wav files and creates `prompts-original` works fine, but there is an issue with the file `gender.txt` created by the script: The contents are different from the ones in `spk2gender` in the git repo:

1. `gender.txt` contains 159 entries, `spk2gender` 162 entries
2. Some entries are only in `gender.txt` and some only in `spk2gender`

Please find attached both files. Can you explain the differences between the files?
[gender.txt](https://github.com/gooofy/speech/files/1942080/gender.txt)
[spk2gender.txt](https://github.com/gooofy/speech/files/1942082/spk2gender.txt)

UPDATE: Sorry, my bad! I know why the files are different: For development purposes I've only taken a sample form the gspv2 corpus to cut down on execution time of the scripts.

mpuels commented:
**Convert gspv2 to VoxForge format: More speakers in gender.txt than in spk2gender**

In the meantime I've run the script `gspv2_to_vf.py` on the whole gspv2 corpus and have compared the generated file `gender.txt` with the file `spk2gender` in the git repo. Now the stats are as follows:

- number of lines in `gender.txt`: 179
- number of lines in `spk2gender`: 162

So now there are considerably more lines in `gender.txt` than in `spk2gender`. And in one instance the gender is wrong in `gender.txt` but correct in `spk2gender`. But why are there so much more entries in `gender.txt`? I've checked one recording of one of the speakers in `gender.txt`, which is not in `spk2gender`. The uttered sentence matches the prompt.

@gooofy Can you explain, why there are more entries in `gender.txt`? Probably we should stick with `spk2gender`, because it contains the corrected gender for one speaker. But what about the missing entries? Should I write them into `spk2gender`?

[spk2gender.txt](https://github.com/gooofy/speech/files/1943144/spk2gender.txt)
[gender.txt](https://github.com/gooofy/speech/files/1943137/gender.txt)

mpuels committed:
Add gspv2_to_vf.py: Script to convert gspv2 corpus to VoxForge format.

mpuels committed:
gspv2_to_vf.py: Get rid of folders dev, test, and train.

gooofy commented:
@mpuels: excellent work (even though that functional coding style frightens me ;) )

> In the meantime I've run the script gspv2_to_vf.py on the whole gspv2 corpus and have compared the generated file gender.txt with the file spk2gender in the git repo. Now the stats are as follows:
> 
> number of lines in gender.txt: 179
> number of lines in spk2gender: 162
> So now there are considerably more lines in gender.txt than in spk2gender. And in one instance the gender is wrong in gender.txt but correct in spk2gender. But why are there so much more entries in gender.txt? I've checked one recording of one of the speakers in gender.txt, which is not in spk2gender. The uttered sentence matches the prompt.

looking at the diff it seems that gender.txt is a true superset of spk2gender, i.e. all speakers in spk2gender appear in gender.txt, right?

investigating this further i appears 18 speakers are missing:

+gsp22bab8af0aa64491bd4dedc01570c646_2015 m
+gsp3386f5a025464a3c9b67880379011ce2_2015 f
+gsp3dbe86f2d94c4e14b270330d31e1c34d_2015 m
+gsp43889a1c06d245409e459bd9ea9cfb67_2015 m
+gsp4a831451c0c244e7b4254467307a29e6_2015 m
+gsp5908a8651bd8447da249aea7c7a6c2b2_2014 f
+gsp5a4078f166df45a582b8b9d17b1bc1c2_2015 f
+gsp61825372365f4595bae0e56d3a288cb3_2015 m
+gsp9718be3a912c4a9d95c72f56f1531c34_2015 m
+gsp9e6a00c980f0479d8b364139a9571217_2015 m
+gspaac053d6fbdc42d998d2592e084aea91_2015 f
+gspb88bc430a1e8416eb112cffd38314500_2015 m
+gspc83081438f824344bd9c1ccb5d6d99fd_2015 m
+gspd2474655a436438abc63204dc23c1034_2015 m
+gspe861ed18097a4e78af6118bf2a842b83_2015 m
+gspea218e91c87448e18d78f09eefe21271_2015 m
+gspfc21e5d132b74adc9f4e29ae0ad8ff85_2015 f
+gspff56a22112fb4ffe9d58bfb7531f8b9d_2015 m

all of them except one (gsp5908a8651bd8447da249aea7c7a6c2b2_2014 - I have no idea why that one is missing) belong to the test set - I must confess that I don't really remember but it could be the case that I simply never got round to reviewing the test set (even the dev/train set review is not finished (6115 files are yet to be reviewed, if I am correct)) so simply never added that to the database. but that is just a guess. 

there is definitely nothing wrong with adding all of them to the db.

mpuels committed:
gspv2 to voxforge: Change name format of speaker id and drop dev,test,train

The old speaker id contained the year of the recording, e.g.
gsp01deac0b92bd44ed8a89ceceaf5da3fc_2015. The new speaker id doesn't have
the year, e.g. gsp01deac0b92bd44ed8a89ceceaf5da3fc.

The function speech_audio_scan.py::scan_audio_dir() expects, that the name
of each subdirectory contains a hyphen. To meet that requirement, the
script gsvp2_to_vf.py now appends "-1" to each speaker id and takes the
resulting string as the name of the subdirectory containing the speaker's
recordings. Example of a name of a transformed subdirectory:
gsp01deac0b92bd44ed8a89ceceaf5da3fc-1.

Finally, the transformed corpus doesn't contain the subdirectories dev,
test, and train anymore.

mpuels commented:
@gooofy Ok, I've added the speaker ids I've found to spk2gender and have committed the changes (please see the previous commit).

gooofy commented:
very coo, thank you! :)

don't forget to add your name to the Copyright header of the scripts you modify, we want to give credit where credit is due ;)

mpuels committed:
Rewrite speech_kaldi_export.py

New call signature:

speech_kaldi_export.py model_name dictionary language_model audio_corpus [audio_corpus ...]

mpuels committed:
gspv2_to_vf.py: Exit if destination directory exists.

mpuels committed:
Update README.md: .speechrc, speech corpora, train kaldi model

mpuels committed:
README.md: Add missing step in example to train Kaldi model

mpuels committed:
Update copyright text in modified source files

mpuels committed:
speech_kaldi_export.py: Add argument --prompt_words

The argument was added by Guenter on master and I'm integrating it into my
development branch.

mpuels committed:
dict-de.ipa: Copy newest version from master into development branch

mpuels committed:
Integrate newest transcripts_*.csv into dev branch

Additionally, add corpus 'forschergeist'.

mpuels committed:
Add corpus zamia_de

The file zamia_de_speakers.txt can be removed after this PR has been
approved.

mpuels committed:
README.md: Add corpora Forschergeist and Zamia (German)

mpuels committed:
Remove de/tokenizer_errors.txt

mpuels committed:
Add zamia_de/spk_test.txt

mpuels committed:
Copy de/dict.ipa (commit sha1: 9d044655cd0, 2018-04-28) to dicts/dict-de.ipa

mpuels commented:
@gooofy I've rebased my development branch on master, so that there shouldn't be any merge conflicts left. If you like, you can now test the new version of the scripts by following the steps in the new README.md in the section "NNet3 Models". Also make sure to update your `~/.speechrc` (an updated template is in README.md).

Looking forward to your feedback!

mpuels committed:
Add missing file kaldi-path.sh.template

gooofy commented:
impressive work! - I have decided to merge this right away, thanks again for all you contributions! (...and keep them coming !)
